### PR TITLE
fix: delete pending pod when workflow terminated 

### PR DESCRIPTION
**Pull request by shuangkun**
When I submit a workflow. It didn't work successed for some reason(such as no stock), some of pod is  still Pending.
So I terminate the workflow.  But the pending pod was not deleted.

After awhile, the pending pod is running , but the workflow has failed and caused some waste of resources.

So, we need to delete pending pod when workflow terminated.

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -512,7 +512,18 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 		pods := wfc.kubeclientset.CoreV1().Pods(namespace)
 		switch action {
 		case terminateContainers:
-			if terminationGracePeriod, err := wfc.signalContainers(namespace, podName, syscall.SIGTERM); err != nil {
+			pod, err := wfc.getPod(namespace, podName)
+			if err == nil && pod != nil && pod.Status.Phase == "Pending" {
+				logCtx.Info("deleting pending pod")
```
Comment (Joibel) [inline workflow/controller/controller.go:517]: Please could you add `.WithFields(log.Fields{"name": podName, "namespace": namespace})` to the logging.

**Review by Joibel — CHANGES_REQUESTED** [review]
I think I agree with the overall intention here, it seems reasonable.

It's a gap which isn't caught by any existing mechanism.

File: workflow/controller/controller.go (hunk 1 above)
Comment (shuangkun) [reply workflow/controller/controller.go:517]: ok，I will add it.

**Review by Joibel — APPROVED** [review]
Looks good, thank you for adding this. You'll need to fix the linter error before this can be merged though.

File: workflow/controller/controller.go [hunk 2]
```diff
@@ -512,7 +512,18 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 		pods := wfc.kubeclientset.CoreV1().Pods(namespace)
 		switch action {
 		case terminateContainers:
-			if terminationGracePeriod, err := wfc.signalContainers(namespace, podName, syscall.SIGTERM); err != nil {
+			pod, err := wfc.getPod(namespace, podName)
+			if err == nil && pod != nil && pod.Status.Phase == "Pending" {
+				logCtx.WithFields(log.Fields{"name": podName, "namespace": namespace}).Info("deleting pending pod")
+				propagation := metav1.DeletePropagationBackground
+				err = pods.Delete(ctx, podName, metav1.DeleteOptions{
```
Comment (sarabala1979) [inline workflow/controller/controller.go:519]: Can we use the Pod cleanup queue delete it?

**Comment by sarabala1979** [comment]
Thanks @Joibel for reviewing.

File: workflow/controller/controller.go (hunk 2 above)
Comment (Joibel) [reply workflow/controller/controller.go:519]: Ah, not something I've come across yet. It sounds cleaner.

Comment (shuangkun) [reply workflow/controller/controller.go]: > Can we use the Pod cleanup queue to delete it?
OK, I have requeue it to cleanup queue to delete it

**Comment by shuangkun** [comment]
> Looks good, thank you for adding this. You'll need to fix the linter error before this can be merged though.



> Looks good, thank you for adding this. You'll need to fix the linter error before this can be merged though.

Do you know how should i do to fix the linter error? I really didn't chang something about it.

**Review by Joibel — APPROVED** [review]
Lovely, thank @shuangkun

**Comment by sarabala1979** [comment]
@shuangkun can you add tests (unit + e2e) ?

**Comment by shuangkun** [comment]
> @shuangkun can you add tests (unit + e2e) ?

I have tested it in my environment，I will try to add some unit tests. This scene is slightly more difficult to construct.

**Comment by shuangkun** [comment]
> @shuangkun can you add tests (unit + e2e) ?

I have add a ut. Pealease have a check.

**Comment by Joibel** [comment]
LGTM, thanks @shuangkun. @sarabala1979, could you re-review and consider merging

File: workflow/controller/controller.go [hunk 3]
```diff
@@ -525,7 +525,10 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 		pods := wfc.kubeclientset.CoreV1().Pods(namespace)
 		switch action {
 		case terminateContainers:
-			if terminationGracePeriod, err := wfc.signalContainers(namespace, podName, syscall.SIGTERM); err != nil {
+			pod, err := wfc.getPod(namespace, podName)
+			if err == nil && pod != nil && pod.Status.Phase == "Pending" {
```
Comment (terrytangyuan) [inline workflow/controller/controller.go:529]: Use the constant from common.go?

Comment (shuangkun) [reply workflow/controller/controller.go]: OK, I use the constant.

Comment (shuangkun) [reply workflow/controller/controller.go]: > Use the constant from common.go?

Could you have a check, thanks!

Comment (Joibel) [reply workflow/controller/controller.go]: LGTM. PTAL @terrytangyuan

**Review by sarabala1979 — APPROVED** [review]
Thanks for your contribution. LGTM. Please update branch

**Pull request merged**